### PR TITLE
nightly check for Bert

### PR DESCRIPTION
**Pull request by samhumeau**
Adding a nightly test for Bert, train for about 100 samples on convai2 for bi and cross ranker.

I wasn't sure how to keep the final evaluation of the train loop under a certain number of samples, 
so I created this debug LimitedSelfOriginalTeacher, that can be used for such occasions.

### CODE REVIEW

**Comment by samhumeau** [comment]
Following instructions from dataset test: 
```
python tests/datatests/test_new_tasks.py
.
----------------------------------------------------------------------
Ran 1 test in 194.066s

OK

```

**Comment by samhumeau** [comment]
So it remains to be seen if it can run on the GPU that take care of the test.
On my machine where I have access to a 16GB GPU, nvidia-smi tells me that this test uses 13GB, but I am not sure how much of that is actually used.

File: tests/nightly/gpu/test_bert.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            warmup_updates=100,
+        ))
+        # The cross encoder reaches an interesting state MUCH faster
+        # accuracy should be present and somewhere between 0.2 and 0.8
+        # (large interval so that it doesn't flake.)
+        self.assertGreaterEqual(
+            test['accuracy'], 0.2,
+            'test accuracy = {}\nLOG:\n{}'.format(test['accuracy'], stdout)
+        )
+        self.assertLessEqual(
```
Comment (stephenroller) [inline tests/nightly/gpu/test_bert.py]: I think we can get rid of the lessEqual constraints. Mostly I think we can just ensure this doesn't crash.

File: parlai/tasks/convai2/agents.py [hunk 2]
```diff
@@ -69,6 +69,18 @@ def __init__(self, opt, shared=None):
         super().__init__(opt, shared)
 
 
+class LimitedSelfOriginalTeacher(SelfOriginalTeacher):
```
Comment (stephenroller) [inline parlai/tasks/convai2/agents.py]: can we just do --num-epochs .001 instead of making a new teacher?

**Comment by stephenroller** [comment]
How long does the test take?

If you add "- nightly_gpu_tests" here, the test will run on this branch:

'https://github.com/facebookresearch/ParlAI/blob/b412c4d88298a7c531eaab9b5e105ab5588f3a87/.circleci/config.yml#L176

File: tests/nightly/gpu/test_bert.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+import unittest
+import parlai.core.testing_utils as testing_utils
+
+
+@testing_utils.skipUnlessGPU
+class TestBertModel(unittest.TestCase):
+    """Checks that Both Biencoder and CrossEncoder of Bert can be trained
```
Comment (stephenroller) [inline tests/nightly/gpu/test_bert.py]: Do you mind editing the docstring whitespace to be more like:

https://github.com/facebookresearch/ParlAI/blob/b412c4d88298a7c531eaab9b5e105ab5588f3a87/parlai/core/torch_agent.py#L1468-L1474

File: parlai/tasks/convai2/agents.py (hunk 2 above)
Comment (samhumeau) [reply parlai/tasks/convai2/agents.py]: This does not seem to apply for the final eval of the train loop (on valid and test set) does it?

File: tests/nightly/gpu/test_bert.py (hunk 1 above)
Comment (samhumeau) [reply tests/nightly/gpu/test_bert.py]: I'd leave it just to make sure no bug produces an accuracy of 1.0, but maybe that's an overstretch

**Comment by samhumeau** [comment]
The tests took less than a minute on my machine

File: parlai/tasks/convai2/agents.py (hunk 2 above)
Comment (stephenroller) [reply parlai/tasks/convai2/agents.py]: Oh jeez the damn validation; you're right.

Comment (stephenroller) [reply parlai/tasks/convai2/agents.py]: I’m going to make a flag to help us with that. I ran into this exact issue two weeks ago

File: tests/nightly/gpu/test_bert.py (hunk 1 above)
Comment (stephenroller) [reply tests/nightly/gpu/test_bert.py]: oh good idea

File: .circleci/config.yml [hunk 4]
```diff
@@ -129,6 +137,7 @@ jobs:
       - setup
       - installtorchgpu
       - installdeps
+      - installhfbert
```
Comment (stephenroller) [inline .circleci/config.yml]: maybe consider renaming this to "installgpudeps" or something

File: parlai/tasks/convai2/agents.py (hunk 2 above)
Comment (stephenroller) [reply parlai/tasks/convai2/agents.py]: cf #1606

**Pull request merged**